feat(caching): Store action outcomes only once

hrushikeshm-g opened the pull request:
Refactor caching mechanism to eliminate redundant outcome storage on the client side, storing outcomes once on the worker side in the permanent cache.

Previously, action outcomes were redundantly stored in:
- `propose-action` cache (cache 1)
- `get_outcome` temporary cache (cache 2, via `*_ref_key`)
- `finalized_episode` temporary cache (cache 3, via `*_ref_key`)
- The completed list.

This new approach streamlines caching by:
- Caching outcomes directly on the worker side using `action_dict_hash` as the key.
- Reducing `finalized_episode` RPC payload size.
- Storing only `outcome_ref_key` in the `completed_list`, which is then transferred to `get_outcome`.

This optimization mandates the use of centralized storage (e.g., GCS or gcs_with_redis), as the completed list now relies on cache keys, making local cache mode incompatible. While slightly increasing implementation complexity, it significantly reduces data redundancy and improves efficiency.

Relates to architectural diagrams "previous_architecture.jpg" and "new_architecture.jpg".

![previous_architecture](https://github.com/user-attachments/assets/7aebb56b-e30e-412b-99b2-b00d2043832c)
![new_architecture](https://github.com/user-attachments/assets/7b1d4461-a8cc-4d7d-831e-28098123df08)

hrushikeshm-g committed:
feat(caching): Store action outcomes only once

Refactor caching mechanism to eliminate redundant outcome storage on the client
side, storing outcomes once on the worker side in the permanent cache.

Previously, action outcomes were redundantly stored in:
- `propose-action` cache (cache 1)
- `get_outcome` temporary cache (cache 2, via `*_ref_key`)
- `finalized_episode` temporary cache (cache 3, via `*_ref_key`)
- The completed list.

This new approach streamlines caching by:
- Caching outcomes directly on the worker side using `action_dict_hash` as the key.
- Reducing `finalized_episode` RPC payload size.
- Storing only `outcome_ref_key` in the `completed_list`, which is then
  transferred to `get_outcome`.

This optimization mandates the use of centralized storage (e.g., GCS or
gcs_with_redis), as the completed list now relies on cache keys, making local
cache mode incompatible. While slightly increasing implementation complexity,
it significantly reduces data redundancy and improves efficiency.

Relates to architectural diagrams "previous_architecture.jpg" and "new_architecture.jpg".

hrushikeshm-g commented on py/sight/utility.py:
```diff
@@ -78,8 +73,10 @@ def get_all_outcomes(sight_id, question_label, action_ids):
             proto=outcome.action_attrs)
-        outcome_dict['outcome'] = convert_proto_to_dict(
-            proto=outcome.outcome_attrs)
+        outcome_dict['outcome'] = json.loads(
+            cache_client.get(outcome.outcome_attrs_ref_key))
```
and otherwise , I should raise the Exception right ? @gemini-code-assist

hrushikeshm-g replied:
@gemini-code-assist , I am updating the as follows

```python
if not outcome_attrs_ref_key:
  raise Exception(f'outcome_attrs_ref_key not found in completed outcome for an {outcome.action_id} due to some issue')
outcome_dict['outcome'] = json.loads(
    cache_client.get(outcome.outcome_attrs_ref_key))
```

hrushikeshm-g replied:
This is better !!

hrushikeshm-g commented on py/sight/proto/sight.proto:
```diff
@@ -761,6 +761,9 @@ message DecisionOutcome {
   float discount = 3;
   DecisionParam outcome_params = 4;
+  // will not be using this in RPC
+  string outcome_params_ref_key = 5;
+  // actual outcome_params_cache_key will be based on action_dict
```
outcome_param_ref_key is the cache_key only , I will update the comment to make it more understandable , works ?  @gemini-code-assist

hrushikeshm-g replied:
updated like below , @gemini-code-assist

```proto
message DecisionOutcome {
  string outcome_label = 1;
  // The numeric value of this outcome, with higher values being more desirable.
  float reward = 2;
  float discount = 3;
  // the below will not be using this in RPC
  DecisionParam outcome_params = 4;
  // actual outcome_params_ref_key will be based on action_dict and we are using
  // this to transfer the outcome via ref_key
  string outcome_params_ref_key = 5;
}

```

hrushikeshm-g replied:
Updated

hrushikeshm-g commented on py/sight/widgets/decision/decision.py:
```diff
@@ -840,13 +843,36 @@ def _update_cached_batch(sight: Any):
+    else:
+      raise Exception(
+          'You are not using any centrailzed storage to pass the message , check your cache mode , it should be `gcs` , `gcs_with_redis` , `*_with_redis`'
+      )
```
updated

hrushikeshm-g commented on sight_service/proto/service.proto:
```diff
@@ -334,14 +334,15 @@ message GetOutcomeResponse {
     Status status = 6;
     string response_str = 7;
     int64 action_id = 8;
+    // sight.x.proto.DecisionParam outcome_attrs = 4; // removing this
+    // completing as completelist will only containt the cache_keys now
+    string outcome_attrs_ref_key = 9;
```
what extra details should I add to explain that ?

@gemini-code-assist

hrushikeshm-g replied:
updated

hrushikeshm-g committed:
fix: Address review comments

hrushikeshm-g commented on py/sight/utility.py:
```diff
@@ -78,8 +73,27 @@ def get_all_outcomes(sight_id, question_label, action_ids):
+          )
+          raise ValueError(
+              f"Outcome data not found in cache for key: {outcome_ref_key}")
+        outcome_dict['outcome'] = json.loads(cached_outcome_data)
```
like what ? , @gemini-code-assist

hrushikeshm-g replied:
updated

hrushikeshm-g commented on py/sight/widgets/decision/decision.py:
```diff
@@ -840,13 +843,36 @@ def _update_cached_batch(sight: Any):
+          discount=sight.widget_decision_state['discount'],
+          reward=sight.widget_decision_state.get('sum_reward', 0),
+          outcome_ref_key=cache_key)
+    else:
+      raise Exception(
+          'You are not using any centralized storage to pass the message. Centralized storage (GCS or Redis) is required for this configuration. Check your cache mode; it should be `gcs`, `gcs_with_redis`, or `*_with_redis`.'
+      )
```
like what ? , @gemini-code-assist

hrushikeshm-g replied:
updated

hrushikeshm-g commented on sight_service/proto/service.proto:
```diff
@@ -334,14 +334,18 @@ message GetOutcomeResponse {
     Status status = 6;
     string response_str = 7;
     int64 action_id = 8;
+    // sight.x.proto.DecisionParam outcome_attrs = 4; // Removed to centralize
+    // outcome storage. The actual outcome attributes are now stored on the
+    // worker side in a permanent cache. `outcome_attrs_ref_key` is used here as
+    // a reference key to these attributes, reducing payload size and ensuring
+    // the completed list only contains cache keys.
+    string outcome_attrs_ref_key = 9;
```
like what ? , @gemini-code-assist

hrushikeshm-g replied:
already added

hrushikeshm-g committed:
fix: Address gemini review comments

bronevet-abc commented on py/sight/demo/proposal_demo.py:
```diff
@@ -47,17 +47,27 @@ def main(argv: Sequence[str]) -> None:
   # create sight object with configuration to spawn workers beforehand
   with Sight.create(params, config) as sight:
 
+    import uuid
```
move imports to the top of the file

hrushikeshm-g replied:
updated

bronevet-abc commented on py/sight/proto/sight.proto:
```diff
@@ -750,7 +750,7 @@ message DecisionPoint {
   string chosen_option = 2; // need or not?
 
-  // The parameters that characterize the chosen option.
+  // The parameters that characterize the chosen option.ssss
```
remove "ssss"

hrushikeshm-g replied:
updated

bronevet-abc commented on py/sight/utility.py:
```diff
@@ -109,6 +126,8 @@ def poll_network_batch_outcome(sight_id, question_label):
         outcome_of_action_ids = get_all_outcomes(sight_id, question_label,
                                                  pending_action_ids)
 
+        # outcome_of_action_ids =
```
can we remove this?

hrushikeshm-g replied:
updated

bronevet-abc commented on py/sight/utility.py:
```diff
@@ -78,8 +73,30 @@ def get_all_outcomes(sight_id, question_label, action_ids):
             proto=outcome.action_attrs)
-        outcome_dict['outcome'] = convert_proto_to_dict(
-            proto=outcome.outcome_attrs)
+        outcome_ref_key = outcome.outcome_attrs_ref_key
```
Please add a comment saying that we're picking up from the cache the outcome object that was written into the cache by the worker.

hrushikeshm-g replied:
updated with 

```python
# The `outcome_ref_key` is crucial for retrieving the full outcome data from the
# centralized cache, as per the new caching mechanism. Its absence indicates a
# critical data flow issue or misconfiguration.
# here we are trying to pick up the data that is stored by worker with given outcome_ref_key as cache_key in centralized cache


```

bronevet-abc commented on py/sight/widgets/decision/decision.py:
```diff
@@ -840,13 +843,38 @@ def _update_cached_batch(sight: Any):
+      logging.info('action_dict used by worker is => %s', action_dict)
+
+      key_maker = KeyMaker()
+      worker_version = 'v1'
```
Why is the worker_version defaulting to v1? Shouldn't it come from our yml config?

hrushikeshm-g replied:
Yes, there was no question label, so for testing, I was using v1 as the default. I need to propagate the question label to call `worker_version = utils.get_worker_version(question_label)`

updating

hrushikeshm-g replied:
updated

bronevet-abc commented on py/sight/widgets/decision/decision.py:
```diff
@@ -840,13 +843,38 @@ def _update_cached_batch(sight: Any):
+          discount=sight.widget_decision_state['discount'],
+          reward=sight.widget_decision_state.get('sum_reward', 0),
+          outcome_ref_key=cache_key)
+    else:
```
This error message is confusing. _update_cached_batch is called in decision_outcome(), so it looks like we've just turned off all support for none, local and local_with_redis modes. Is that the case?

Won't the mechanism still work if the workers have access to the redis instance on the local machine of the client?

hrushikeshm-g replied:
No , the system only works when there is one centralized caching mode like `gcs` , `gcs_with_redis`  or `*_with_redis` where redis is centralized

hrushikeshm-g replied:
@gemini-code-assist , can you try to explain code flow to greg , why do we need centralized cache

bronevet-abc replied:
I like the new explanation mechanism :)

The problem I was having is that the error message was saying that only these modes are allowed but then why was the user even allowed to select local and none caches?

I think the key thing is that we're using redis to manage communication between clients and workers and to make this happen we need a single redis to be shared between the two. We need a way to to keep a temporary cache for experiments (e.g. tests), which should be achievable via a local redis instance that can be spun up for a single run or be flushed on demand. To this end can we remove local and none modes (make them not even be selectable) but keep a mode where a local redis instance is used instead of the managed MemoryStore instance?

hrushikeshm-g replied:
Yes, updating the `cache_mode` flags to `['gcs', 'redis', 'gcs_with_redis']` will work for us, so `local` and `none` are not even selectable.

works ? @gemini-code-assist

hrushikeshm-g replied:
Update _CACHE_MODE flag , checkout commit :  https://github.com/theteamatx/x-sight/pull/98/commits/63a067666f024d655d7a16c992d84e54aa4f392e

hrushikeshm-g committed:
fix: gregs comments are solved

hrushikeshm-g committed:
fix: fixed the test_decision.py

hrushikeshm-g committed:
fix: Address the comment for _CACHE_MODE

update the _CACHE_MODE flag to select only three types of mode which are
of following types

- gcs
- redis
- gcs_with_redis

hrushikeshm-g committed:
fix: move the test_decision to integraton dir

hrushikeshm-g committed:
fix: redis constant updated

hrushikeshm-g committed:
fix: address the test-cases issue

hrushikeshm-g committed:
fix: remove the unwated file grep

hrushikeshm-g committed:
fix: Redis init the config based on the flags

hrushikeshm-g committed:
Merge branch 'dev' into feature/cache-once
